### PR TITLE
GEOMESA-2524 Hand-walk FSDS metadata directory

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Using FileContext.utils() is orders of magnitudes slower with large
  numbers of files
* Fix USE_PROVIDED_FID behavior for FSDS FeatureStore
* Add docs on setting user.timezone for spark sql

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

File: docs/user/filesystem/sql.rst [hunk 1]
```diff
@@ -79,7 +82,6 @@ have ingested GDELT data into S3 you can query it with SQL in the spark shell::
 
     // Select the top event codes
     spark.sql("SELECT eventCode, count(*) as count FROM gdelt " +
-              "WHERE dtg >= cast('2017-06-01T00:00:00.000' as timestamp) " +
-              "AND dtg <= cast('2017-06-30T00:00:00.000' as timestamp) " +
+              "WHERE dtg >= '2017-06-01T00:00:00Z' AND dtg <= '2017-06-30T00:00:00Z' " +
```
Comment (jnh5y) [inline docs/user/filesystem/sql.rst]: you got rid of the cast?!

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/MetadataFileSystemStorage.scala [hunk 2]
```diff
@@ -127,48 +127,52 @@ abstract class MetadataFileSystemStorage(conf: Configuration,
   override def compact(partition: String, threads: Int): Unit = {
     val toCompact = getFilePaths(partition).asScala
 
-    val leaf = metadata.getPartitionScheme.isLeafStorage
-    val dataPath = StorageUtils.nextFile(metadata.getRoot, partition, leaf, extension, FileType.Compacted)
+    if (toCompact.lengthCompare(2) < 0) {
+      logger.debug(s"Skipping compaction for single data file: ${toCompact.mkString(", ")}")
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/MetadataFileSystemStorage.scala]: Hmmm... seems like we wouldn't need the .mkString for one file:)?

File: docs/user/filesystem/sql.rst (hunk 1 above)
Comment (elahrvivaz) [reply docs/user/filesystem/sql.rst]: yeah it's not needed any more - we fixed that a while ago. it still works with the cast but it seems cumbersome

File: geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/compact/CompactCommand.scala [hunk 3]
```diff
@@ -60,21 +61,21 @@ class CompactCommand extends FsDataStoreCommand with LazyLogging {
...
+
+    // compact metadata up-front, that will save us listing redundant metadata files for each partition
+    storage.getMetadata.compact()
+
     mode match {
       case RunModes.Local =>
-        toCompact.map(_.name).foreach { p =>
-          logger.info(s"Compacting $p")
-          storage.compact(p)
+        toCompact.par.foreach { p =>
```
Comment (jnh5y) [inline geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/compact/CompactCommand.scala]: Do we have control over the number of threads used by this .par?

File: geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/MetadataFileSystemStorage.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/MetadataFileSystemStorage.scala]: could also be 0 files... it was either that or `toCompact.headOption.map(_.toString).getOrElse("")`

File: geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/compact/CompactCommand.scala (hunk 3 above)
Comment (elahrvivaz) [reply geomesa-fs/geomesa-fs-tools/src/main/scala/org/locationtech/geomesa/fs/tools/compact/CompactCommand.scala]: you can control it through various scala-level settings - it defaults to numProcessors I believe. But I've cleaned this up in a follow-on PR that exposes num-threads as a parameter: https://github.com/locationtech/geomesa/pull/2176/files#diff-acb9f51edcb091704a605d5517f01b45R140

**Pull request merged**